### PR TITLE
🌱 Set build variables during `make build`

### LINKED ISSUE #1401: Build-related flags for executables built from source
# Description

Currently, when `kubebuilder` is built from source, `kubebuilder version` yields:
```
Version: version.Version{KubeBuilderVersion:"unknown", KubernetesVendor:"unknown", GitCommit:"$Format:%H$", BuildDate:"1970-01-01T00:00:00Z", GoOs:"unknown", GoArch:"unknown"}
```

These variables are set as part of the automated build process, but they are never stored in the release branch. they can be found in https://github.com/kubernetes-sigs/kubebuilder/blob/9552d129f7d12cc6c7072a1d323ed9f4b461d550/cmd/version/version.go#L27-L35

# Feature request

Set these flags as part of the development process instead so that, if kubebuilder is built from a release branch, even if it is from source, this fields will be set to provide useful information.

There are three types of variables:

 1. `kubeBuilderVersion`: should be set in the branch.
 2. Those that should be set in the build process:
    1. `goos`: can't be stored in the release branch as there is no one branch per target.
    2. `goarch`: same.
    3. `gitCommit`: the git commit sha1 can't be obtained before commiting the release, so it needs to be set afterwards.
    4. `buildDate`: obvious reasons.
 3. `kubernetesVendorVersion`: has no meaning when building from source.

/kind feature
/kind documentation

### PR DESCRIPTION
**Pull request by Adirio**
# Description
Several variables are set when releasing a kubebuilder version in order to output the correct information when calling `kubebuilder version`. But when building from the source with `make build`, this variables were not being set, and thus the `kubebuilder version` output was providing no information:
```
Version: version.Version{KubeBuilderVersion:"unknown", KubernetesVendor:"unknown", GitCommit:"$Format:%H$", BuildDate:"1970-01-01T00:00:00Z", GoOs:"unknown", GoArch:"unknown"}
```

This PR updates those variables to reflect the current working environment.

Fixes #1401

P.S.: it also cleans a bit the `Makefile`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -41,18 +41,22 @@ help:  ## Display this help
 
 .PHONY: build
 build: ## Build the project locally
-	go build -o bin/kubebuilder ./cmd
+	BUILD_VERSION=$(shell git descripe --tags --dirty --broken)
+	BUILD_OS=$(shell go env GOOS)
+	BUILD_ARCH=$(shell go env GOARCH)
+	BUILD_COMMIT=$(shell git rev-parse HEAD)
+	BUILD_DATE=$(shell date -u +'%Y-%m-%dT%H:%M:%SZ')
+	LD_FLAGS=-ldflags "-X sigs.k8s.io/kubebuilder/v2/cmd.kubeBuilderVersion=$(BUILD_VERSION) -X sigs.k8s.io/kubebuilder/v2/cmd.goos=$(BUILD_OS) -X sigs.k8s.io/kubebuilder/v2/cmd.goarch=$(BUILD_ARCH) -X sigs.k8s.io/kubebuilder/v2/cmd.gitCommit=$(BUILD_COMMIT)-X sigs.k8s.io/kubebuilder/v2/cmd.buildDate=$(BUILD_DATE)"
```
Comment (estroz) [inline Makefile:49]:
```suggestion
```

File: Makefile [hunk 2]
```diff
@@ -67,15 +71,15 @@ lint-fix: golangci-lint ## Run golangci lint to automatically perform fixes
 	@$(GOLANGCI_LINT) run --fix
 
 golangci-lint:
```
Comment (estroz) [inline Makefile:73]: This target is modified significantly by #1813, thoughts?

File: Makefile [hunk 3]
```diff
@@ -22,9 +22,9 @@ CONTROLLER_GEN_BIN_PATH := $(shell which controller-gen)
 
 # Get the currently used golang install path (in GOPATH/bin, unless GOBIN is set)
 ifeq (,$(shell go env GOBIN))
-GOBIN=$(shell go env GOPATH)/bin
+	GOBIN=$(shell go env GOPATH)/bin
```
Comment (estroz) [inline Makefile:25]: Idiomatic Makefiles define variables without indentation to denote that they are always global, so can you revert these changes?

**Review by estroz — CHANGES_REQUESTED** [review]
Suggestions

File: Makefile (hunk 3 above)
Comment (Adirio) [reply Makefile:25]: It was to differentiate easier which one is saved under which conditions, but I'll revert it.

**Comment by Adirio** [comment]
@estroz all suggestions applied

**Review by estroz — APPROVED** [review]
/lgtm

**Comment by Adirio** [comment]
@estroz would you mind re-applying `lgtm` label as it had merge conflicts due to #1813 which also modified this file and had to rebase it and resolve them.
